engine/src/main/java/org/hibernate/validator/internal/engine/valueextraction/ValueExtractorResolver.java

```diff
@@ -0,0 +1,431 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.internal.engine.valueextraction;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import javax.validation.ConstraintDeclarationException;
+import javax.validation.valueextraction.ValueExtractor;
+
+import org.hibernate.validator.internal.metadata.aggregated.CascadingMetaDataBuilder;
+import org.hibernate.validator.internal.metadata.aggregated.ContainerCascadingMetaData;
+import org.hibernate.validator.internal.metadata.aggregated.PotentiallyContainerCascadingMetaData;
+import org.hibernate.validator.internal.util.CollectionHelper;
+import org.hibernate.validator.internal.util.Contracts;
+import org.hibernate.validator.internal.util.ReflectionHelper;
+import org.hibernate.validator.internal.util.TypeHelper;
+import org.hibernate.validator.internal.util.TypeVariableBindings;
+import org.hibernate.validator.internal.util.TypeVariables;
+import org.hibernate.validator.internal.util.logging.Log;
+import org.hibernate.validator.internal.util.logging.LoggerFactory;
+import org.hibernate.validator.internal.util.stereotypes.Immutable;
+
+/**
+ * Contains resolving algorithms for {@link ValueExtractor}s, and caches for these
+ * extractors based on container types.
+ *
+ * @author Gunnar Morling
+ * @author Guillaume Smet
+ * @author Marko Bekhta
+ */
+public class ValueExtractorResolver {
+
+	private static final Log LOG = LoggerFactory.make( MethodHandles.lookup() );
+
+	private static final Object NON_CONTAINER_VALUE = new Object();
+
+	@Immutable
+	private final Set<ValueExtractorDescriptor> registeredValueExtractors;
+
+	private final ConcurrentHashMap<ValueExtractorCacheKey, Set<ValueExtractorDescriptor>> possibleValueExtractorsByRuntimeTypeAndTypeParameter = new ConcurrentHashMap<>();
+
+	private final ConcurrentHashMap<Class<?>, Set<ValueExtractorDescriptor>> possibleValueExtractorsByRuntimeType = new ConcurrentHashMap<>();
+
+	private final ConcurrentHashMap<Class<?>, Object> nonContainerTypes = new ConcurrentHashMap<>();
```

Comment:
Should this be using `ConcurrentReferenceHashMap` instead? The key of `Class` makes me a bit nervous in terms of deployment classes leaking.

Reply:
Since it's not static, what are you concerned about exactly?

But also: since the key is a Class, there's a hyper officient specialized map implementation for such cases in the JDK. I forgot the name :( but it was originally introduced to optimize "invoke dynamic".

Reply:
My concern is about a `ValidatorFactory` managed by the container outliving the deployment. Probably you're right, though, and this should not happen in practice. FWIW, we use CRHM in other places where we keep metadata by class.

On that hyper-efficient impl, above we're instantiating `CHM` directly, so we wouldn't benefit from that? Or are you saying that CHM is internally using an optimized path if keys are class objects?

Reply:
It's a different collection container, not sure it implements the Map or the ConcurrentMap API but it's essentially an equivalent of Map<Class,X>. I wish I could remember its name, as I'm needing to try it out in another context too :-/